src/test/scala/com/workshop/RollingWindowThrottlerTest.scala

```diff
@@ -48,35 +51,26 @@ class RollingWindowThrottler(
                               durationWindow: FiniteDuration,
                               clock: Clock) {
 
-  val invocations = scala.collection.mutable.HashMap.empty[String, Invocation]
+  val invocations : LoadingCache[String, Counter] = CacheBuilder.newBuilder()
+        .expireAfterWrite(durationWindow.toMillis, TimeUnit.MILLISECONDS)
+        .ticker(throttlerTicker())
+        .build(defaultCounter())
 
   def tryAcquire(key: String): Try[Unit] = {
-    val invocation = invocations.getOrElseUpdate(key, newInvocation)
-    (expirationHandler orElse throttlerHandler)(key -> invocation)
-  }
-
-  val expirationHandler: PartialFunction[(String, Invocation), Try[Unit]] = {
-    case (key, invocation) if expires(invocation) => {
-      invocations -= key
-      tryAcquire(key)
-    }
-  }
-
-  val throttlerHandler: PartialFunction[(String, Invocation), Try[Unit]] = {
-    case (key, invocation) => if (invocation.counter.incrementAndGet <= max) {
+    if(invocations.get(key).incrementAndGet <= max){
       Success()
-    } else {
+    }else {
       Failure(new ThrottleException)
     }
   }
 
-  def newInvocation: Invocation =
-    Invocation(clock.instant(), Counter())
-
-
-  def expires(invocation: Invocation): Boolean =
-    (clock.instant().toEpochMilli - invocation.timestamp.toEpochMilli) >= durationWindow.toMillis
+  def defaultCounter(): CacheLoader[String, Counter] = new CacheLoader[String, Counter] {
+    override def load(key: String): Counter = Counter()
+  }
 
+  def throttlerTicker(): Ticker = new Ticker {
+    override def read(): Long = TimeUnit.MILLISECONDS.toNanos(clock.instant().toEpochMilli)
```

Comment:
isn't this a bit complex? not saying it's wrong but I just remember my impl was shorter